Make transpose use SIMD

nilsdeppe opened the pull request:
## Proposed changes

This speeds up the transpose and partial derivatives by significantly reducing memory accesses during transposition. Partial derivatives runtime on an 8^3 grid is reduced by 33%, so a 1.5x speedup.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Make transpose use SIMD

This speeds up the transpose and partial derivatives by significantly reducing
memory accesses during transposition.

wthrowe commented:
I'll trust that this implementation is the right way to do it.  The test shows that it works.

I think you've managed to completely avoid ODR-using the AVX/SSE2 specializations of `transpose_block` when those features are not supported.  If so, you could just `#if` the entire functions out instead of replacing the body with an error.  That should also reduce the number of warnings somewhat.  (I'm not going to require this PR to be clang-tidy clean, but easy improvements are nice.)

nilsdeppe commented:
Agreed! I rebased and changed the macro locations so that the warning push/pop isn't needed. I also added a `NOLINTBEGIN`, which was added in clang-14 (I think CI is not that new), so at least in the future these will go away. I figure add a `NOLINTNEXTLINE` would be quite tedious both for reading and writing since half the lines would have that...

wthrowe commented:
Ignoring clang-tidy as discussed above.
